### DISCUSSION TITLE
Decoders

### DISCUSSION THREAD

**Discussion by AntoineLaporte** [pschatzmann/arduino-audio-tools#652] [Q&A] [answered]
Hi,
I'm trying to send read some files through Bluetooth and the sound is not good unless I'm reading 16bit files.
I tried WAV and MP3, from SD or SPIFFS and it sounds always the same like if the decoding has some latency. I guess it's not the BT because it's working fine with the 16bit files.
I'm not a very good programmer but I'm always keep to learn.
The goal of the program it to send numbers to earphones. At the end it should be integrated in an other program I'm already using and the files would be on SPIFFS because I'm already writing continuously on the SD.
I've looked for the last 3 days all around and I did not find something about that kind of problem.
Thanks for any help !

```
#include <Arduino.h>
//remove this, to find issues regarding mp3 decoding
#define HELIX_LOGGING_ACTIVE false

#include "AudioTools.h"
#include "AudioLibs/AudioA2DP.h"
#include "AudioLibs/AudioSourceSD.h"

#include "AudioCodecs/CodecMP3Helix.h"
//#include "AudioLibs/AudioKit.h"

#define PIN_AUDIO_KIT_SD_CARD_CS 5

const char *startFilePath="/";
const char* ext="mp3";
AudioSourceSD source(startFilePath, ext, PIN_AUDIO_KIT_SD_CARD_CS);
A2DPStream bt;
MP3DecoderHelix decoder;
AudioPlayer player(source, bt, decoder);

bool pathSet = false;
bool firstPlay = false;
bool secondPlay = false;
long timer = 0;
long timer2 = 0;


void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);

  // Setting up SPI if necessary with the right SD pins by calling 
  // SPI.begin(PIN_AUDIO_KIT_SD_CARD_CLK, PIN_AUDIO_KIT_SD_CARD_MISO, PIN_AUDIO_KIT_SD_CARD_MOSI, PIN_AUDIO_KIT_SD_CARD_CS);

  // setup output - We send the test signal via A2DP - so we conect to a Bluetooth Speaker
  auto cfg = bt.defaultConfig(TX_MODE);
  //cfg.name = "iLuv Mobi";  // set the device here. Otherwise the next available device is used for output
  //cfg.auto_reconnect = true;  // if this is use we just quickly connect to the last device ignoring cfg.name
  bt.begin(cfg);

  // setup player
  player.setVolume(1);
  player.begin();
  bool pathSet = false;
  timer = millis()-5000;
  timer2 = millis();
  player.setPath("/test.mp3");
}

void loop() {
  if ((millis()-timer) < 5000 && !player.isActive()){
    if (!firstPlay) {
      player.setPath("/OneHundred.mp3");
      player.play();
      firstPlay = true;
    }
    else if (!secondPlay) {
      player.setPath("/Seventeen.mp3");
      player.play();
      secondPlay = true;
    }
  }
  if ((millis()-timer) > 5000){
    firstPlay = false;
    secondPlay = false;
    timer = millis();
    delay(1000);
  }
  if ((millis()-timer2)>2000)
    player.copy();
  //Serial.println(player.isActive());
}
```

**Comment by pschatzmann** [comment, marked as answer]
I don't quite understand your problem.  But to solve your task, why don't you just use this: https://github.com/pschatzmann/arduino-simple-tts. Or you can get some inspiration from this how to solve your problem.
But whatever you do be advised that A2DP has always a big latency (around 1 sec)!

**Reply by AntoineLaporte** [reply]
My problem is the audio quality when I play an mp3, it's really bad with some lag.
I will try the TTS for sure !

**Comment by AntoineLaporte** [comment]
I'm actually trying the TTS and I have an other problem, the audio is getting out in something that sounds like double speed.
I'm using a Bluetooth speaker.
I think problem is the rate that is 44100 for A2P and 24000 in the TTS example sketch ("cfg.sample_rate = 24000"), but I cannot do this command with an A2DPConfig object. Do I understand it well and what would be the way around ?

**Reply by pschatzmann** [reply]
I suggest that you read the chapter about resampling and converting the number of channels in the Wiki.
To save memory it is advisable to store the data in a lower sample rate. If you need to provide it in another rate you can just resample.

**Comment by AntoineLaporte** [comment]
I've managed to resample, but I still have the same problem of sound.
When I look to the Serial I got this and I can see the lines coming one after the other, as the part of the "full sound" are coming one after the over, what I called "lag" before. Like "Two--En--Ty--------Se--Ven"
So what I'm understanding is that the StreamCopy is not fast enough.

```
[I] NumberToText.h : 35 - say(number='27', decimal='')
[I] TextToSpeech.h : 58 - say: TWENTY
[I] AudioCopy.h : 77 - buffer_size=1024
[I] AudioTypes.h : 63 - sample_rate: 24000
[I] AudioTypes.h : 64 - channels: 1
[I] AudioTypes.h : 65 - bits_per_sample: 16
[I] AudioCopy.h : 134 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 134 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 134 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 134 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 134 - StreamCopy::copy 224 -> 224 -> 224 bytes - in 1 hops
[I] TextToSpeech.h : 58 - say: SEVEN
[I] AudioCopy.h : 77 - buffer_size=1024
[I] AudioCopy.h : 134 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 134 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 134 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[I] AudioCopy.h : 134 - StreamCopy::copy 864 -> 864 -> 864 bytes - in 1 hops


```

**Reply by pschatzmann** [reply]
Yes of cause you get a pause if you do it like this !
Provide all words which need to be spoken together as a Vector and this should resolve the issue. 
Finally also set the Log level to Warning will help with the audio quality as well.
The TextToSpeechQueue might also help...

**Reply by AntoineLaporte** [reply]
I will find out the Vector thing and the TTSQueue too.
Thank you for your help !

**Reply by AntoineLaporte** [reply]
I've read around and found the Queue, I used the example with TTS and I still get the same result.
For the Vector it's probably above my actual knowledge.
I'm gonna give up for now on using BT for this project, I'm not sure the ESP32 can handle reading-parsing the GPS data, logging on SD and sending sound to BT all at the same time. I'm gonna wait for some DAC board and test with plugged headphones.
Anyway it was very interesting to try to do it, I've learnt some new things !

**Reply by pschatzmann** [reply]
To test you can also try to use the AnalogAudioStream together with some earphones...
